primitives/src/slot.rs

```diff
@@ -0,0 +1,773 @@
+use std::iter::FromIterator;
+use std::collections::BTreeMap;
+use std::fmt;
+
+use bitvec::prelude::BitVec;
+use bitvec::cursor::BigEndian;
+use itertools::Itertools;
+
+use beserial::{Deserialize, Serialize, ReadBytesExt, WriteBytesExt, SerializingError, SerializeWithLength, DeserializeWithLength, uvar};
+use bls::bls12_381::lazy::LazyPublicKey;
+use bls::bls12_381::CompressedPublicKey;
+use keys::Address;
+
+use crate::coin::Coin;
+use crate::policy::SLOTS;
+
+
+/// Enum to index a slot.
+///
+/// You can either address a slot directly by its slot number, or by it's band number. The band
+/// number than corresponds to either the nth [StakeBand] or [ValidatorBand], depending on which
+/// collection it is applied.
+///
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SlotIndex {
+    Slot(u16),
+    Band(u16),
+}
+
+
+/// A slot that contains the corresponding validator and stake information.
+///
+/// # ToDo
+///
+/// * We could include the slot number in here
+///
+#[derive(Clone, Debug)]
+pub struct Slot {
+    pub validator_slot: ValidatorSlotBand,
+    pub stake_slot: StakeSlotBand,
+}
+
+impl Slot {
+    pub fn public_key(&self) -> &LazyPublicKey {
+        self.validator_slot.public_key()
+    }
+
+    pub fn staker_address(&self) -> &Address {
+        self.stake_slot.staker_address()
+    }
+
+    pub fn reward_address(&self) -> &Address {
+        self.stake_slot.reward_address()
+    }
+}
+
+/// Complete view of all slots
+///
+/// # ToDo
+///
+/// * Can we use references to the [ValidatorSlots] and [StakeSlots]? If so, we can just reference
+///   them right from the block. Maybe something like *SuperCow* works here?
+///
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
+pub struct Slots {
+    pub validator_slots: ValidatorSlots,
+    pub stake_slots: StakeSlots,
+}
+
+impl Slots {
+    pub fn new(validator_slots: ValidatorSlots, stake_slots: StakeSlots) -> Self {
+        Slots {
+            validator_slots,
+            stake_slots,
+        }
+    }
+
+    pub fn get(&self, idx: SlotIndex) -> Option<Slot> {
+        Some(Slot {
+            validator_slot: self.validator_slots.get(idx)?.clone(),
+            stake_slot: self.stake_slots.get(idx)?.clone(),
+        })
+    }
+
+    #[deprecated]
+    pub fn slash_fine(&self) -> Coin {
+        Coin::ZERO
+    }
+
+    /// Returns stake slots with first slot number and the corresponding public key of the
+    /// validator slot.
+    pub fn combined(&self) -> Vec<(Slot, u16)> {
+        let mut combined = Vec::new();
+        let mut validator_slots_iter = self.validator_slots.iter().peekable();
+        let mut start_stake_slot_number = 0;
+        let mut start_validator_slot_number = 0;
+
+        for stake_slot in self.stake_slots.iter() {
+            let validator_slot = *validator_slots_iter.peek()
+                .expect("Expected validator slot");
+
+            let end_stake_slot_number = start_stake_slot_number + stake_slot.num_slots();
```

Comment:
I would move these two declarations down to the place where they are actually needed.
Makes it a bit more readable IMO.